File: EventDrivenStreamingPlatform/infrastructure/src/main/java/com/netflix/streaming/infrastructure/exception/GlobalExceptionHandler.java
package com.netflix.streaming.infrastructure.exception;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.validation.FieldError;
import org.springframework.web.bind.MethodArgumentNotValidException;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.bind.annotation.RestControllerAdvice;
import org.springframework.web.context.request.WebRequest;

import java.time.Instant;
import java.util.HashMap;
import java.util.Map;
import java.util.UUID;

/**
 * Global exception handler for consistent error responses across all services.
 * 
 * Provides:
 * - Consistent error response format
 * - Proper HTTP status codes
 * - Error correlation IDs for tracing
 * - Security-conscious error messages
 */
@RestControllerAdvice
public class GlobalExceptionHandler {

    private static final Logger logger = LoggerFactory.getLogger(GlobalExceptionHandler.class);

    @ExceptionHandler(MethodArgumentNotValidException.class)
    public ResponseEntity<ErrorResponse> handleValidationException(
            MethodArgumentNotValidException ex, WebRequest request) {
        
        Map<String, String> errors = new HashMap<>();
        ex.getBindingResult().getFieldErrors().forEach(error -> {
            errors.put(error.getField(), error.getDefaultMessage());
        });

        String errorId = UUID.randomUUID().toString();
        logger.warn("Validation failed [errorId={}, path={}]", errorId, request.getDescription(false));

        ErrorResponse error = ErrorResponse.builder()
                .errorId(errorId)
                .timestamp(Instant.now())
                .status(HttpStatus.BAD_REQUEST.value())
                .error("Validation Failed")
                .message("Request validation failed")
                .path(request.getDescription(false).replace("uri=", ""))
                .details(errors)
                .build();

        return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(error);
    }

    @ExceptionHandler(IllegalArgumentException.class)
    public ResponseEntity<ErrorResponse> handleIllegalArgumentException(
            IllegalArgumentException ex, WebRequest request) {
        
        String errorId = UUID.randomUUID().toString();
        logger.warn("Illegal argument [errorId={}, message={}]", errorId, ex.getMessage());

        ErrorResponse error = ErrorResponse.builder()
                .errorId(errorId)
                .timestamp(Instant.now())
                .status(HttpStatus.BAD_REQUEST.value())
                .error("Bad Request")
                .message(ex.getMessage())
                .path(request.getDescription(false).replace("uri=", ""))
                .build();

        return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(error);
    }

    @ExceptionHandler(IllegalStateException.class)
    public ResponseEntity<ErrorResponse> handleIllegalStateException(
            IllegalStateException ex, WebRequest request) {
        
        String errorId = UUID.randomUUID().toString();
        logger.warn("Illegal state [errorId={}, message={}]", errorId, ex.getMessage());

        ErrorResponse error = ErrorResponse.builder()
                .errorId(errorId)
                .timestamp(Instant.now())
                .status(HttpStatus.CONFLICT.value())
                .error("Conflict")
                .message(ex.getMessage())
                .path(request.getDescription(false).replace("uri=", ""))
                .build();

        return ResponseEntity.status(HttpStatus.CONFLICT).body(error);
    }

    @ExceptionHandler(ResourceNotFoundException.class)
    public ResponseEntity<ErrorResponse> handleResourceNotFoundException(
            ResourceNotFoundException ex, WebRequest request) {
        
        String errorId = UUID.randomUUID().toString();
        logger.debug("Resource not found [errorId={}, resource={}]", errorId, ex.getResourceType());

        ErrorResponse error = ErrorResponse.builder()
                .errorId(errorId)
                .timestamp(Instant.now())
                .status(HttpStatus.NOT_FOUND.value())
                .error("Not Found")
                .message(ex.getMessage())
                .path(request.getDescription(false).replace("uri=", ""))
                .build();

        return ResponseEntity.status(HttpStatus.NOT_FOUND).body(error);
    }

    @ExceptionHandler(EventPublishingException.class)
    public ResponseEntity<ErrorResponse> handleEventPublishingException(
            EventPublishingException ex, WebRequest request) {
        
        String errorId = UUID.randomUUID().toString();
        logger.error("Event publishing failed [errorId={}]", errorId, ex);

        ErrorResponse error = ErrorResponse.builder()
                .errorId(errorId)
                .timestamp(Instant.now())
                .status(HttpStatus.INTERNAL_SERVER_ERROR.value())
                .error("Event Publishing Failed")
                .message("Failed to publish event. Please retry.")
                .path(request.getDescription(false).replace("uri=", ""))
                .build();

        return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body(error);
    }

    @ExceptionHandler(Exception.class)
    public ResponseEntity<ErrorResponse> handleGenericException(
            Exception ex, WebRequest request) {
        
        String errorId = UUID.randomUUID().toString();
        logger.error("Unexpected error [errorId={}]", errorId, ex);

        ErrorResponse error = ErrorResponse.builder()
                .errorId(errorId)
                .timestamp(Instant.now())
                .status(HttpStatus.INTERNAL_SERVER_ERROR.value())
                .error("Internal Server Error")
                .message("An unexpected error occurred. Please contact support with error ID: " + errorId)
                .path(request.getDescription(false).replace("uri=", ""))
                .build();

        return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body(error);
    }

    /**
     * Standardized error response structure
     */
    public static class ErrorResponse {
        private String errorId;
        private Instant timestamp;
        private int status;
        private String error;
        private String message;
        private String path;
        private Map<String, Object> details;

        private ErrorResponse(Builder builder) {
            this.errorId = builder.errorId;
            this.timestamp = builder.timestamp;
            this.status = builder.status;
            this.error = builder.error;
            this.message = builder.message;
            this.path = builder.path;
            this.details = builder.details;
        }

        public static Builder builder() {
            return new Builder();
        }

        public static class Builder {
            private String errorId;
            private Instant timestamp;
            private int status;
            private String error;
            private String message;
            private String path;
            private Map<String, Object> details;

            public Builder errorId(String errorId) {
                this.errorId = errorId;
                return this;
            }

            public Builder timestamp(Instant timestamp) {
                this.timestamp = timestamp;
                return this;
            }

            public Builder status(int status) {
                this.status = status;
                return this;
            }

            public Builder error(String error) {
                this.error = error;
                return this;
            }

            public Builder message(String message) {
                this.message = message;
                return this;
            }

            public Builder path(String path) {
                this.path = path;
                return this;
            }

            public Builder details(Map<String, Object> details) {
                this.details = details;
                return this;
            }

            public ErrorResponse build() {
                return new ErrorResponse(this);
            }
        }

        // Getters
        public String getErrorId() { return errorId; }
        public Instant getTimestamp() { return timestamp; }
        public int getStatus() { return status; }
        public String getError() { return error; }
        public String getMessage() { return message; }
        public String getPath() { return path; }
        public Map<String, Object> getDetails() { return details; }
    }
}

File: KotlinPaymentsPlatform/shared/src/main/kotlin/com/example/kotlinpay/shared/exception/GlobalExceptionHandler.kt
package com.example.kotlinpay.shared.exception

import org.slf4j.LoggerFactory
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.validation.FieldError
import org.springframework.web.bind.MethodArgumentNotValidException
import org.springframework.web.bind.annotation.ExceptionHandler
import org.springframework.web.bind.annotation.RestControllerAdvice
import org.springframework.web.context.request.WebRequest
import java.time.Instant
import java.util.UUID

/**
 * Global exception handler for consistent error responses across all payment services.
 * 
 * Provides:
 * - Consistent error response format
 * - Proper HTTP status codes
 * - Error correlation IDs for tracing
 * - Security-conscious error messages (no sensitive data exposure)
 */
@RestControllerAdvice
class GlobalExceptionHandler {

    private val logger = LoggerFactory.getLogger(GlobalExceptionHandler::class.java)

    @ExceptionHandler(MethodArgumentNotValidException::class)
    fun handleValidationException(
        ex: MethodArgumentNotValidException,
        request: WebRequest
    ): ResponseEntity<ErrorResponse> {
        val errors = ex.bindingResult.fieldErrors.associate { 
            it.field to (it.defaultMessage ?: "Invalid value")
        }

        val errorId = UUID.randomUUID().toString()
        logger.warn("Validation failed [errorId={}, path={}]", errorId, request.getDescription(false))

        val error = ErrorResponse(
            errorId = errorId,
            timestamp = Instant.now(),
            status = HttpStatus.BAD_REQUEST.value(),
            error = "Validation Failed",
            message = "Request validation failed",
            path = request.getDescription(false).replace("uri=", ""),
            details = errors
        )

        return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(error)
    }

    @ExceptionHandler(IllegalArgumentException::class)
    fun handleIllegalArgumentException(
        ex: IllegalArgumentException,
        request: WebRequest
    ): ResponseEntity<ErrorResponse> {
        val errorId = UUID.randomUUID().toString()
        logger.warn("Illegal argument [errorId={}, message={}]", errorId, ex.message)

        val error = ErrorResponse(
            errorId = errorId,
            timestamp = Instant.now(),
            status = HttpStatus.BAD_REQUEST.value(),
            error = "Bad Request",
            message = ex.message ?: "Invalid request",
            path = request.getDescription(false).replace("uri=", "")
        )

        return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(error)
    }

    @ExceptionHandler(IllegalStateException::class)
    fun handleIllegalStateException(
        ex: IllegalStateException,
        request: WebRequest
    ): ResponseEntity<ErrorResponse> {
        val errorId = UUID.randomUUID().toString()
        logger.warn("Illegal state [errorId={}, message={}]", errorId, ex.message)

        val error = ErrorResponse(
            errorId = errorId,
            timestamp = Instant.now(),
            status = HttpStatus.CONFLICT.value(),
            error = "Conflict",
            message = ex.message ?: "Invalid state",
            path = request.getDescription(false).replace("uri=", "")
        )

        return ResponseEntity.status(HttpStatus.CONFLICT).body(error)
    }

    @ExceptionHandler(ResourceNotFoundException::class)
    fun handleResourceNotFoundException(
        ex: ResourceNotFoundException,
        request: WebRequest
    ): ResponseEntity<ErrorResponse> {
        val errorId = UUID.randomUUID().toString()
        logger.debug("Resource not found [errorId={}, resource={}]", errorId, ex.resourceType)

        val error = ErrorResponse(
            errorId = errorId,
            timestamp = Instant.now(),
            status = HttpStatus.NOT_FOUND.value(),
            error = "Not Found",
            message = ex.message ?: "Resource not found",
            path = request.getDescription(false).replace("uri=", "")
        )

        return ResponseEntity.status(HttpStatus.NOT_FOUND).body(error)
    }

    @ExceptionHandler(PaymentProcessingException::class)
    fun handlePaymentProcessingException(
        ex: PaymentProcessingException,
        request: WebRequest
    ): ResponseEntity<ErrorResponse> {
        val errorId = UUID.randomUUID().toString()
        logger.error("Payment processing failed [errorId={}, code={}]", errorId, ex.errorCode, ex)

        val error = ErrorResponse(
            errorId = errorId,
            timestamp = Instant.now(),
            status = HttpStatus.INTERNAL_SERVER_ERROR.value(),
            error = "Payment Processing Failed",
            message = "Payment processing failed. Please retry or contact support.",
            path = request.getDescription(false).replace("uri=", ""),
            errorCode = ex.errorCode
        )

        return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body(error)
    }

    @ExceptionHandler(SecurityException::class)
    fun handleSecurityException(
        ex: SecurityException,
        request: WebRequest
    ): ResponseEntity<ErrorResponse> {
        val errorId = UUID.randomUUID().toString()
        logger.warn("Security violation [errorId={}]", errorId, ex)

        val error = ErrorResponse(
            errorId = errorId,
            timestamp = Instant.now(),
            status = HttpStatus.FORBIDDEN.value(),
            error = "Security Violation",
            message = "Access denied",
            path = request.getDescription(false).replace("uri=", "")
        )

        return ResponseEntity.status(HttpStatus.FORBIDDEN).body(error)
    }

    @ExceptionHandler(Exception::class)
    fun handleGenericException(
        ex: Exception,
        request: WebRequest
    ): ResponseEntity<ErrorResponse> {
        val errorId = UUID.randomUUID().toString()
        logger.error("Unexpected error [errorId={}]", errorId, ex)

        val error = ErrorResponse(
            errorId = errorId,
            timestamp = Instant.now(),
            status = HttpStatus.INTERNAL_SERVER_ERROR.value(),
            error = "Internal Server Error",
            message = "An unexpected error occurred. Please contact support with error ID: $errorId",
            path = request.getDescription(false).replace("uri=", "")
        )

        return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body(error)
    }

    /**
     * Standardized error response structure
     */
    data class ErrorResponse(
        val errorId: String,
        val timestamp: Instant,
        val status: Int,
        val error: String,
        val message: String,
        val path: String,
        val details: Map<String, Any>? = null,
        val errorCode: String? = null
    )
}

/**
 * Custom exceptions for payment platform
 */
class ResourceNotFoundException(
    val resourceType: String,
    val resourceId: String,
    message: String = "$resourceType with id '$resourceId' not found"
) : RuntimeException(message)

class PaymentProcessingException(
    val errorCode: String,
    message: String,
    cause: Throwable? = null
) : RuntimeException(message, cause)

